reactive/kotlinx-coroutines-reactor/src/Mono.kt

```diff
@@ -67,17 +69,16 @@ private class MonoCoroutine<in T>(
     }
 
     override fun onCancelled(cause: Throwable, handled: Boolean) {
-        try {
-            /*
-             * sink.error handles exceptions on its own and, by default, handling of undeliverable exceptions is a no-op.
-             * Guard potentially non-empty handlers against meaningless cancellation exceptions
-             */
-            if (getCancellationException() !== cause) {
+        /** Cancellation exceptions that were caused by [dispose], that is, came from downstream, are not errors. */
+        if (getCancellationException() !== cause || !disposed) {
```

Comment:
With such a check, we have a scenario when an exception is swallowed.
E.g. the following test:
```
@Test
fun testSwallowedException() = runTest {
    val started = CountDownLatch(1)
    val disposed = CountDownLatch(1)
    val m = mono {
        started.countDown()
        disposed.await()
        println("Running")
        throw ArithmeticException()
    }
    val d = m.subscribe()
    started.await()
    d.dispose()
    disposed.countDown()
}
```

will print `Running` and then will completely ignore the arithmetic exception

Reply:
PTAL on discussion here https://github.com/Kotlin/kotlinx.coroutines/issues/252 and the corresponding commit that fixes the problem

Reply:
I have read the discussion and am not sure how it is applicable. Could you please clarify what behavior would you expect instead?

What happens here is that `onError` gets called after `m` is disposed of, and Reactor deals with error from such objects by passing them to the `onOperatorError` hook, which is unset in this test. Setting the hook leads to the exception being reported:
```kotlin
    @Test
    fun testSwallowedException() = runTest {
        Hooks.onOperatorError("testSwallowedException") { t, a ->
            expectUnreached()
            t
        }
        val started = CountDownLatch(1)
        val disposed = CountDownLatch(1)
        val m = mono {
            started.countDown()
            disposed.await()
            println("Running")
            throw ArithmeticException()
        }
        val d = m.subscribe()
        started.await()
        d.dispose()
        disposed.countDown()
    }
```
successfully fails.

Moreover, it looks like using `onOperatorError` is the approach we take for undeliverable exceptions in Reactor, so… this exception is already dealt with the way we usually deal with the undeliverable exceptions. What am I missing here?

Reply:
Oh, sorry, my bad, I've misinterpreted the code. Thanks for the clarification!